### PR TITLE
DOC: Add note to docstring DataFrame.compare about identical labels

### LINKED ISSUE #35491: DOC: Add note to DataFrame.compare about only able to compare identically-labeled DataFrame objects
#### Location of the documentation
`pandas.core.frame::compare`

#### Documentation problem
Right now there is not a note/warning that you can only compare dataframes with the same shape and index labels.

#### Suggested fix for documentation
Add note: 

> Note: the method can only compare identically-labeled DataFrame objects, this means DataFrames with identical row and column labels.

### PR DESCRIPTION
**Pull request by erfannariman**
- [x] closes #35491
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by erfannariman** [comment]
@simonjayhawkins do you have time to check this PR? Thanks

**Comment by simonjayhawkins** [comment]
will look soon

File: pandas/core/frame.py [hunk 1]
```diff
@@ -5899,6 +5899,10 @@ def _construct_result(self, result) -> "DataFrame":
 -----
 Matching NaNs will not appear as a difference.
 
+Can only compare identically-labeled DataFrame objects,
+this means DataFrames with the same shape and
+identical row, column labels.
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:5904]: if the user tries this they will get a `ValueError: Can only compare identically-labeled DataFrame objects` 

so can you also add a Raises section to the docstring.

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
Thanks @erfannariman for the PR. generally lgtm. ex @MarcoGorelli comment and explicit mention of `ValueError` in a raises section.

File: pandas/core/frame.py (hunk 1 above)
Comment (erfannariman) [reply pandas/core/frame.py]: Yes, makes sense, added.

[the remark this answers is not in the document]
Comment (erfannariman) [reply pandas/core/frame.py]: Yes, done.

**Comment by erfannariman** [comment]
@MarcoGorelli I added `DataFrame.equals` to "see also"  section as well since these methods are quite alike. If that is not oke, let me know and I'll remove it.

File: pandas/core/frame.py [hunk 2]
```diff
@@ -5891,17 +5891,22 @@ def _construct_result(self, result) -> "DataFrame":
     The resulting index will be a MultiIndex with 'self' and 'other'
     stacked alternately at the inner level.
 
+Raises
+------
+ValueError
+    * When `errors='raise'` and both DataFrames don't have identical labels.
```
Comment (simonjayhawkins) [inline pandas/core/frame.py:5897]: compare doesn't have a errors parameter. doesn't this raise ValueError always if both DataFrames don't have identical labels?

Comment (erfannariman) [reply pandas/core/frame.py]: True, made an edit, hopefully I understood you correctly.

**Review by simonjayhawkins — APPROVED** [review]
Thanks @erfannariman lgtm pending green

**Comment by erfannariman** [comment]
@MarcoGorelli Do you have any idea what's the cause of this test failing? I can't figure out.

**Comment by simonjayhawkins** [comment]
> @MarcoGorelli Do you have any idea what's the cause of this test failing? I can't figure out.

fixed in #36011, restarted job so should pass.

**Comment by erfannariman** [comment]
ping @MarcoGorelli @simonjayhawkins

**Review by simonjayhawkins — APPROVED** [review]

File: pandas/core/frame.py [hunk 3]
```diff
@@ -5932,14 +5932,23 @@ def _construct_result(self, result) -> "DataFrame":
     The resulting index will be a MultiIndex with 'self' and 'other'
     stacked alternately at the inner level.
 
+Raises
+------
+ValueError
+    * When both DataFrames don't have identical labels or shape.
```
Comment (MarcoGorelli) [inline pandas/core/frame.py:5938]: First, the formatting looks a bit different to other docstrings, e.g. in pandas/core/algorithms.py we find

```
    Raises
    ------
    IndexError
        When `indices` is out of bounds for the array.
    ValueError
        When the indexer contains negative values other than ``-1``
        and `allow_fill` is True.
```

, is the bullet point `* ` like this used elsewhere in the codebase?

Second, I think the usage of "both" here isn't usual, how about:

```
ValueError
    When the two DataFrames don't have identical labels or shape.
```

File: pandas/core/frame.py [hunk 4]
```diff
@@ -5932,14 +5932,23 @@ def _construct_result(self, result) -> "DataFrame":
...
 --------
 Series.compare : Compare with another Series and show differences.
+DataFrame.equals : Test whether two objects contain the same elements.
 
 Notes
 -----
 Matching NaNs will not appear as a difference.
 
+Can only compare identically-labeled DataFrame objects,
+this means DataFrames with the same shape and identical row, column labels.
```
Comment (MarcoGorelli) [inline pandas/core/frame.py:5950]: Suggestion (which might need splitting across multiple lines):

```
Can only compare identically-labeled (i.e. same shape, identical row and column labels) DataFrames
```

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
Thanks! Another couple of comments, should be good after this

File: pandas/core/frame.py (hunk 3 above)
Comment (erfannariman) [reply pandas/core/frame.py]: Yes, first I had two raises, so had bullet points, forgot to remove later. Adjusted this.

**Review by MarcoGorelli — APPROVED** [review]
Great, thanks @erfannariman !

**Comment by jbrockmendel** [comment]
LGTM

**Comment by MarcoGorelli** [comment]
Thanks @erfannariman

**Pull request merged**